engine-transactions/src/eip_7702.rs

```diff
@@ -0,0 +1,262 @@
+use crate::eip_2930::AccessTuple;
+use crate::Error;
+use aurora_engine_precompiles::secp256k1::ecrecover;
+use aurora_engine_types::types::{Address, Wei};
+use aurora_engine_types::{Vec, H160, U256};
+use evm::executor::stack::Authorization;
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// Type indicator (per EIP-7702)
+pub const TYPE_BYTE: u8 = 0x04;
+
+// EIP-7702 `MAGIC` number
+pub const MAGIC: u8 = 0x5;
+
+/// The order of the secp256k1 curve, divided by two. Signatures that should be checked according
+/// to EIP-2 should have an S value less than or equal to this.
+///
+/// `57896044618658097711785492504343953926418782139537452191302581570759080747168`
+pub const SECP256K1N_HALF: U256 = U256([
+    0xDFE9_2F46_681B_20A0,
+    0x5D57_6E73_57A4_501D,
+    0xFFFF_FFFF_FFFF_FFFF,
+    0x7FFF_FFFF_FFFF_FFFF,
+]);
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct AuthorizationTuple {
+    pub chain_id: U256,
+    pub address: H160,
+    pub nonce: u64,
+    pub parity: U256,
+    pub r: U256,
+    pub s: U256,
+}
+
+impl Decodable for AuthorizationTuple {
+    fn decode(rlp: &Rlp<'_>) -> Result<Self, DecoderError> {
+        let chain_id = rlp.val_at(0)?;
+        let address = rlp.val_at(1)?;
+        let nonce = rlp.val_at(2)?;
+        let parity = rlp.val_at(3)?;
+        let r = rlp.val_at(4)?;
+        let s = rlp.val_at(5)?;
+        Ok(Self {
+            chain_id,
+            address,
+            nonce,
+            parity,
+            r,
+            s,
+        })
+    }
+}
+
+/// EIP-7702 transaction kind from the Prague hard fork.
+///
+/// See [EIP-7702](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7702.md)
+/// for more details.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct Transaction7702 {
+    /// ID of chain which the transaction belongs.
+    pub chain_id: u64,
+    /// A monotonically increasing transaction counter for this sender
+    pub nonce: U256,
+    /// Determined by the sender and is optional. Priority Fee is also known as Miner Tip as it is
+    /// paid directly to block producers.
+    pub max_priority_fee_per_gas: U256,
+    /// Maximum amount the sender is willing to pay to get their transaction included in a block.
+    pub max_fee_per_gas: U256,
+    /// The maximum amount of gas the sender is willing to consume on a transaction.
+    pub gas_limit: U256,
+    /// The receiving address (`None` for the zero address).
+    pub to: Option<Address>,
```

Comment:
I think we should model this as simply `pub to: Address`. The EIP-7702 documentation clearly states `a null destination is not valid`, so we should not allow `to: None`. If a user wants `to` equal to the zero address then they must explicitly set `to: 0x00`.